File: opencode/tmux-manager/src/tmux.ts
export interface TmuxResult {
  stdout: string;
  exitCode: number;
}

export function runTmuxRaw(args: string[]): TmuxResult {
  const proc = Bun.spawnSync(['tmux', ...args], {
    stdout: 'pipe',
    stderr: 'pipe',
  });
  return {
    stdout: new TextDecoder().decode(proc.stdout).trimEnd(),
    exitCode: proc.exitCode,
  };
}

export function runTmux(args: string[]): string {
  const result = runTmuxRaw(args);
  if (result.exitCode !== 0) {
    throw new Error(`tmux ${args.join(' ')} failed with code ${result.exitCode}`);
  }
  return result.stdout;
}

export function getGlobalOption(name: string, defaultValue: string): string {
  const result = runTmuxRaw(['show-option', '-gqv', name]);
  return result.stdout || defaultValue;
}

export function getSessionOption(session: string, name: string): string | null {
  const result = runTmuxRaw(['show-options', '-qv', '-t', session, name]);
  return result.stdout || null;
}

export function setSessionOption(session: string, name: string, value: string): void {
  runTmux(['set-option', '-t', session, name, value]);
}

export function setGlobalOption(name: string, value: string): void {
  runTmux(['set-option', '-g', name, value]);
}

/** Single-quote a string for safe interpolation into a shell command. */
export function shellQuote(s: string): string {
  return `'${s.replace(/'/g, "'\\''")}'`;
}

/**
 * Build a `tmux attach-session` command string.
 * Pass clearTmuxEnv when the command runs inside another tmux client (e.g. a
 * preview split) so tmux doesn't refuse with "sessions should be nested with
 * care".
 */
export function attachCommand(name: string, opts?: { clearTmuxEnv?: boolean }): string {
  const env = opts?.clearTmuxEnv ? 'env -u TMUX ' : '';
  return `${env}tmux attach-session -t ${shellQuote(name)}`;
}

export interface ClientInfo {
  client: string;
  session: string;
}

export function listClients(): ClientInfo[] {
  const result = runTmuxRaw(['list-clients', '-F', '#{client_name} #{session_name}']);
  if (result.exitCode !== 0 || !result.stdout) return [];
  return result.stdout.split('\n').map((line) => {
    const [client, ...sessionParts] = line.split(' ');
    return { client, session: sessionParts.join(' ') };
  });
}

export function hasSession(name: string): boolean {
  return runTmuxRaw(['has-session', '-t', name]).exitCode === 0;
}

export function newSession(name: string, cwd: string, command: string): void {
  runTmux(['new-session', '-d', '-s', name, '-c', cwd, command]);
}

export function killSession(name: string): void {
  runTmux(['kill-session', '-t', name]);
}

export function detachClient(session: string): void {
  runTmux(['detach-client', '-s', session]);
}

export function displayMessage(format: string, target?: string): string {
  const args = ['display-message', '-p'];
  if (target) args.push('-t', target);
  args.push(format);
  return runTmux(args);
}

export function displayPopup(args: {
  width: string;
  height: string;
  command: string;
  client?: string;
}): void {
  const tmuxArgs = ['display-popup', '-w', args.width, '-h', args.height, '-E', args.command];
  if (args.client) {
    tmuxArgs.splice(1, 0, '-c', args.client);
  }
  runTmux(tmuxArgs);
}

File: llm-session-manager/internal/tmux/tmux.go
package tmux

import (
	"bytes"
	"fmt"
	"os/exec"
	"strings"
)

// Result captures the output and exit code of a tmux invocation.
type Result struct {
	Stdout   string
	ExitCode int
}

// RunRaw runs a tmux command and returns raw output without checking the exit code.
func RunRaw(args []string) Result {
	cmd := exec.Command("tmux", args...)
	var out bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = nil
	err := cmd.Run()
	exitCode := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			exitCode = exitErr.ExitCode()
		} else {
			exitCode = 1
		}
	}
	return Result{Stdout: strings.TrimRight(out.String(), "\n"), ExitCode: exitCode}
}

// Run runs a tmux command and returns its stdout, or an error if it fails.
func Run(args []string) (string, error) {
	result := RunRaw(args)
	if result.ExitCode != 0 {
		return "", fmt.Errorf("tmux %s failed with code %d", strings.Join(args, " "), result.ExitCode)
	}
	return result.Stdout, nil
}

// GetGlobalOption reads a global tmux option, returning defaultValue if unset.
func GetGlobalOption(name, defaultValue string) string {
	result := RunRaw([]string{"show-option", "-gqv", name})
	if result.Stdout == "" {
		return defaultValue
	}
	return result.Stdout
}

// GetSessionOption reads a session option, returning "" if unset.
func GetSessionOption(session, name string) string {
	result := RunRaw([]string{"show-options", "-qv", "-t", session, name})
	return result.Stdout
}

// SetSessionOption sets a session option.
func SetSessionOption(session, name, value string) error {
	_, err := Run([]string{"set-option", "-t", session, name, value})
	return err
}

// SetGlobalOption sets a global option.
func SetGlobalOption(name, value string) error {
	_, err := Run([]string{"set-option", "-g", name, value})
	return err
}

// ShellQuote single-quotes a string for safe interpolation into a shell command.
func ShellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "'\\''") + "'"
}

// AttachCommand builds a tmux attach-session command string.
func AttachCommand(name string, clearTmuxEnv bool) string {
	env := ""
	if clearTmuxEnv {
		env = "env -u TMUX "
	}
	return fmt.Sprintf("%stmux attach-session -t %s", env, ShellQuote(name))
}

// ClientInfo describes an attached tmux client.
type ClientInfo struct {
	Client  string
	Session string
}

// ListClients returns the list of attached clients.
func ListClients() []ClientInfo {
	result := RunRaw([]string{"list-clients", "-F", "#{client_name} #{session_name}"})
	if result.ExitCode != 0 || result.Stdout == "" {
		return nil
	}
	lines := strings.Split(result.Stdout, "\n")
	clients := make([]ClientInfo, 0, len(lines))
	for _, line := range lines {
		parts := strings.SplitN(line, " ", 2)
		if len(parts) != 2 {
			continue
		}
		clients = append(clients, ClientInfo{Client: parts[0], Session: parts[1]})
	}
	return clients
}

// HasSession reports whether a tmux session exists.
func HasSession(name string) bool {
	return RunRaw([]string{"has-session", "-t", name}).ExitCode == 0
}

// NewSession creates a new detached session running command in cwd.
func NewSession(name, cwd, command string) error {
	_, err := Run([]string{"new-session", "-d", "-s", name, "-c", cwd, command})
	return err
}

// KillSession kills a tmux session.
func KillSession(name string) error {
	_, err := Run([]string{"kill-session", "-t", name})
	return err
}

// DetachClient detaches all clients attached to session.
func DetachClient(session string) error {
	_, err := Run([]string{"detach-client", "-s", session})
	return err
}

// DisplayMessage runs display-message with an optional target.
func DisplayMessage(format string, target string) (string, error) {
	args := []string{"display-message", "-p"}
	if target != "" {
		args = append(args, "-t", target)
	}
	args = append(args, format)
	return Run(args)
}

// DisplayPopupOptions configures a tmux display-popup command.
type DisplayPopupOptions struct {
	Width   string
	Height  string
	Command string
	Client  string
}

// DisplayPopup opens a tmux popup.
func DisplayPopup(opts DisplayPopupOptions) error {
	args := []string{"display-popup", "-w", opts.Width, "-h", opts.Height, "-E", opts.Command}
	if opts.Client != "" {
		args = append(args[:1], append([]string{"-c", opts.Client}, args[1:]...)...)
	}
	_, err := Run(args)
	return err
}
